pwndbg/dbg/lldb.py

```diff
@@ -9,22 +13,363 @@
 import pwndbg
 
 
+class LLDBFrame(pwndbg.dbg_mod.Frame):
+    def __init__(self, inner: lldb.SBFrame):
+        self.inner = inner
+
+    @override
+    def evaluate_expression(self, expression: str) -> pwndbg.dbg_mod.Value:
+        value = self.inner.EvaluateExpression(expression)
+        opt_out = _is_optimized_out(value)
+
+        if not value.error.Success() and not opt_out:
+            raise pwndbg.dbg_mod.Error(value.error.description)
+
+        return LLDBValue(value)
+
+
+def map_type_code(type: lldb.SBType) -> pwndbg.dbg_mod.TypeCode:
+    """
+    Determines the type code of a given LLDB SBType.
+    """
+    c = type.GetTypeCode()
+    f = type.GetTypeFlags()
+
+    assert c != lldb.eTypeClassInvalid, "passed eTypeClassInvalid to map_type_code"
+
+    if c == lldb.eTypeClassUnion:
+        return pwndbg.dbg_mod.TypeCode.UNION
+    if c == lldb.eTypeClassStruct:
+        return pwndbg.dbg_mod.TypeCode.STRUCT
+    if c == lldb.eTypeClassTypedef:
+        return pwndbg.dbg_mod.TypeCode.TYPEDEF
+    if c == lldb.eTypeClassPointer:
+        return pwndbg.dbg_mod.TypeCode.POINTER
+    if c == lldb.eTypeClassArray:
+        return pwndbg.dbg_mod.TypeCode.ARRAY
+
+    if f & lldb.eTypeIsInteger != 0:
+        return pwndbg.dbg_mod.TypeCode.INT
+
+    raise RuntimeError("missing mapping for type code")
+
+
+def _is_optimized_out(value: lldb.SBValue) -> bool:
+    """
+    Returns whether the given value is likely to have been optimized out.
+    """
+
+    # We use this rather hacky way to distinguish if expressions that
+    # contain values that have been optimized out, from those that are truly
+    # invalid.
+    #
+    # Obviously, this is a rather bad solution, and breaks if the version of
+    # LLDB we're running under is not in English, or if this message gets
+    # changed in the future.
+    #
+    # LLDB does internally have a way to distinguish the invalid expression
+    # case from the optimized-out one, through lldb::ExpressionResults, but
+    # there does not seem to be a way to wrangle one out of
+    # EvaluateExpression.
```

Comment:
What does this mean that it has a way to distinguish it internally? Is it in its C/C++ code and that is not exposed into Python API? Can we link that code here and inform on that better here?

Reply:
It's part of the private API, and LLDB can distinguish between these cases using it. I've poked around for it and it really seems the public API doesn't have access to it at all.

Reply:
> Can we link that code here and inform on that better here?

I guess we could link the three instances of `EvaluateExpression`? I'm not sure what could be clearer.